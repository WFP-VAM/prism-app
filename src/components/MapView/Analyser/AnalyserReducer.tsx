import { useReducer } from 'react';
import {
  LayerKey,
  AggregationOperations,
  isLayerKey,
  ActionMap,
  isStatistic,
} from '../../../config/types';

export interface AnalyserForm {
  hazardLayerId?: LayerKey;
  baselineLayerId?: LayerKey;
  statistic?: AggregationOperations;
  selectedDate?: number;
  belowThreshold?: string;
  aboveThreshold?: string;
  thresholdError?: string;
}

// Initial form values.
const initForm: AnalyserForm = {
  statistic: AggregationOperations.Mean,
};

export interface URLParams {
  hazardLayerId?: string;
  baselineLayerId?: string;
  statistic?: string;
  selectedDate?: string;
  belowThreshold?: string;
  aboveThreshold?: string;
}

export const URLParamList: string[] = [
  'hazardLayerId',
  'baselineLayerId',
  'statistic',
  'selectedDate',
  'belowThreshold',
  'aboveThreshold',
];

type AnlayserPayload = {
  ['SET_FORM']: { params: URLParams };
  ['SET_LAYER_ID']: {
    type: 'hazard' | 'baseline';
    layerKey: LayerKey;
  };
  ['SET_STATISTIC']: { statistic: AggregationOperations };
  ['SET_SELECTED_DATE']: { date?: number };
  ['SET_THRESHOLD']: {
    type: 'above' | 'below';
    value: string;
  };
  ['CLEAR_FORM']: undefined;
};

export type AnlayserAction = ActionMap<AnlayserPayload>[keyof ActionMap<
  AnlayserPayload
>];

const analyserReducer = (state: AnalyserForm, action: AnlayserAction) => {
  switch (action.type) {
    case 'SET_FORM': {
      const payloadParams: URLParams = action.payload.params;
      const formValues: AnalyserForm = state;

      if (
        payloadParams.hazardLayerId &&
        isLayerKey(payloadParams.hazardLayerId)
      ) {
        // eslint-disable-next-line fp/no-mutation
        formValues.hazardLayerId = payloadParams.hazardLayerId;
      }

      if (
        payloadParams.baselineLayerId &&
        isLayerKey(payloadParams.baselineLayerId)
      ) {
        // eslint-disable-next-line fp/no-mutation
        formValues.baselineLayerId = payloadParams.baselineLayerId;
      }

      if (payloadParams.selectedDate) {
        // eslint-disable-next-line fp/no-mutation
        formValues.selectedDate = parseInt(payloadParams.selectedDate, 10);
      }

      const statisticCorrected: AggregationOperations = isStatistic(
        payloadParams.statistic,
      )
        ? payloadParams.statistic
        : AggregationOperations.Mean;

      if (payloadParams.belowThreshold) {
        // eslint-disable-next-line fp/no-mutation
        formValues.belowThreshold = payloadParams.belowThreshold;
      }

      if (payloadParams.aboveThreshold) {
        // eslint-disable-next-line fp/no-mutation
        formValues.aboveThreshold = payloadParams.aboveThreshold;
      }

      return {
        ...state,
        ...formValues,
        statistic: statisticCorrected,
      };
    }
    case 'SET_LAYER_ID': {
      if (isLayerKey(action.payload.layerKey)) {
        if (action.payload.type === 'hazard') {
          return {
            ...state,
            hazardLayerId: action.payload.layerKey,
          };
        }

        return {
          ...state,
          baselineLayerId: action.payload.layerKey,
        };
      }
      return state;
    }
    case 'SET_STATISTIC': {
      return {
        ...state,
        statistic: action.payload.statistic,
      };
    }
    case 'SET_SELECTED_DATE': {
      return {
        ...state,
        selectedDate: action.payload.date,
      };
    }
    case 'SET_THRESHOLD': {
      const hasError: boolean =
        action.payload.type === 'below'
          ? ((state.aboveThreshold &&
              action.payload.value < state.aboveThreshold) as boolean)
          : ((state.belowThreshold &&
              state.belowThreshold < action.payload.value) as boolean);

      const thresholdErrorString: string = hasError
        ? 'Min threshold is larger than Max!'
        : '';

      if (action.payload.type === 'below') {
        return {
          ...state,
          belowThreshold: action.payload.value,
          thresholdError: thresholdErrorString,
        };
      }

      return {
        ...state,
        aboveThreshold: action.payload.value,
        thresholdError: thresholdErrorString,
      };
    }
    case 'CLEAR_FORM': {
      return initForm;
    }
    default:
      return state;
  }
};

export const useAnalyserReducer = () => {
  return useReducer(analyserReducer, initForm);
};
