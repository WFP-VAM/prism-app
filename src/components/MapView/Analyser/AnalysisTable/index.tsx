import React, { useState } from 'react';
import {
  createStyles,
  Table,
  TableBody,
  TableCell,
  TableContainer,
  TableHead,
  TablePagination,
  TableRow,
  TableSortLabel,
  withStyles,
  WithStyles,
} from '@material-ui/core';
import { orderBy } from 'lodash';
import { useDispatch } from 'react-redux';
import { TableRow as AnalysisTableRow } from '../../../../context/analysisResultStateSlice';
import { showPopup } from '../../../../context/tooltipStateSlice';
import { Column } from '../../../../utils/analysis-utils';

function AnalysisTable({ classes, tableData, columns }: AnalysisTableProps) {
  const [page, setPage] = useState(0);
  const [rowsPerPage, setRowsPerPage] = useState(10);
  const [sortColumn, setSortColumn] = useState<Column['id']>();
  const [isAscending, setIsAscending] = useState(true);

  const dispatch = useDispatch();

  const handleChangePage = (event: unknown, newPage: number) => {
    setPage(newPage);
  };

  const handleChangeRowsPerPage = (
    event: React.ChangeEvent<HTMLInputElement>,
  ) => {
    setRowsPerPage(+event.target.value);
    setPage(0);
  };

  const handleChangeOrderBy = (newSortColumn: Column['id']) => {
    const newIsAsc = !(sortColumn === newSortColumn && isAscending);
    setSortColumn(newSortColumn);
    setIsAscending(newIsAsc);
  };
  return (
    <div>
      <TableContainer className={classes.tableContainer}>
        <Table stickyHeader aria-label="sticky table">
          <TableHead>
            <TableRow>
              {columns.map(column => (
                <TableCell key={column.id} className={classes.tableHead}>
                  <TableSortLabel
                    active={sortColumn === column.id}
                    direction={
                      sortColumn === column.id && !isAscending ? 'desc' : 'asc'
                    }
                    onClick={() => handleChangeOrderBy(column.id)}
                  >
                    {column.label}
                  </TableSortLabel>
                </TableCell>
              ))}
            </TableRow>
          </TableHead>
          <TableBody>
            {orderBy(tableData, sortColumn, isAscending ? 'asc' : 'desc')
              .slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage)
              .map(row => {
                return (
                  <TableRow
                    hover
                    role="checkbox"
                    tabIndex={-1}
                    key={row.name}
                    onClick={() => {
                      // TODO if we decide to keep, add popup data?
                      if (row.coordinates) {
                        dispatch(
                          showPopup({
                            coordinates: row.coordinates,
                            locationName: row.name,
                          }),
                        );
                      }
                    }}
                    style={{ cursor: row.coordinates ? 'pointer' : 'none' }}
                  >
                    {columns.map(column => {
                      const value = row[column.id];
                      return (
                        <TableCell key={column.id}>
                          {column.format && typeof value === 'number'
                            ? column.format(value)
                            : value}
                        </TableCell>
                      );
                    })}
                  </TableRow>
                );
              })}
          </TableBody>
        </Table>
      </TableContainer>
      <TablePagination
        rowsPerPageOptions={[10, 25, 100]}
        component="div"
        count={tableData.length}
        rowsPerPage={rowsPerPage}
        page={page}
        onChangePage={handleChangePage}
        onChangeRowsPerPage={handleChangeRowsPerPage}
      />
    </div>
  );
}

const styles = () =>
  createStyles({
    tableContainer: {
      border: '2px solid',
      width: '700px',
      maxWidth: '100vw',
    },
    tableHead: {
      backgroundColor: '#3d474a',
    },
    innerAnalysisButton: {
      backgroundColor: '#3d474a',
    },
  });

interface AnalysisTableProps extends WithStyles<typeof styles> {
  tableData: AnalysisTableRow[];
  columns: Column[];
}

export default withStyles(styles)(AnalysisTable);
