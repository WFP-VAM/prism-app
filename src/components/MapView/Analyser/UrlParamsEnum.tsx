export enum UrlParamsEnum {
  HAZARD_LAYER = 'hazardLayer',
  STATISTIC = 'statistic',
  BASELINE_LAYER = 'baselineLayer',
  BELOW_THRESHOLD = 'belowThreshold',
  ABOVE_THRESHOLD = 'aboveThreshold',
  DATE = 'date',
}
