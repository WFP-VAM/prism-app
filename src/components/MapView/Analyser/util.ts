import { URLParams } from './AnalyserReducer';

export interface propsFromURL {
  fromURL?: boolean;
  params: URLParams;
}

/**
 * Read parameter from an URL.
 *
 * @param path Browser URL.
 * @param URLParamList List o paramater to be read.
 * @returns Object with all valid URL parameters.
 */
export const extractPropsFromURL = (
  path: string,
  URLParamList: string[],
): propsFromURL => {
  const params: URLSearchParams = new URLSearchParams(path);
  const fromShare: boolean = params.get('share') === 'true';
  const paserdURLParams: URLParams = {};

  // Parse URL only if detected as shared URL.
  if (fromShare) {
    URLParamList.forEach((key: string) => {
      if (params.get(key)) {
        // eslint-disable-next-line fp/no-mutation
        paserdURLParams[key as keyof URLParams] = params.get(key) as string;
      }
    });
  }

  return { fromURL: fromShare, params: paserdURLParams };
};

/**
 * Remove form params from the URL.
 *
 * @param path Browser URL;
 * @param URLParamList List of parameter to be filtered from the URL.
 * @returns New URL;
 */
export const removePropsFromURL = (
  path: string,
  URLParamList: string[],
): string => {
  const params: URLSearchParams = new URLSearchParams(path);
  const keys: IterableIterator<string> = params.keys();
  const urlParams: string[] = [];

  // eslint-disable-next-line no-restricted-syntax
  for (const key of keys) {
    if (!URLParamList.includes(key) && key !== 'share') {
      // eslint-disable-next-line fp/no-mutating-methods
      urlParams.push(`${key}=${params.get(key)}`);
    }
  }

  return urlParams.join('&');
};
