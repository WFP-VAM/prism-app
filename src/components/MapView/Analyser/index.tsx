import React, { useEffect, useMemo, useRef, useState } from 'react';
import {
  Box,
  Button,
  createStyles,
  FormControl,
  FormControlLabel,
  FormGroup,
  Input,
  LinearProgress,
  Radio,
  RadioGroup,
  Switch,
  TextField,
  Theme,
  Typography,
  withStyles,
  WithStyles,
} from '@material-ui/core';
import { grey } from '@material-ui/core/colors';
import { useHistory } from 'react-router-dom';
import { ArrowDropDown, BarChart } from '@material-ui/icons';
import { useDispatch, useSelector } from 'react-redux';
import bbox from '@turf/bbox';
import DatePicker from 'react-datepicker';

import { extractPropsFromURL, propsFromURL, removePropsFromURL } from './util';
import {
  getBoundaryLayerSingleton,
  LayerDefinitions,
} from '../../../config/utils';
import {
  AggregationOperations,
  BoundaryLayerProps,
  NSOLayerProps,
  WMSLayerProps,
  LayerKey,
} from '../../../config/types';
import { LayerData } from '../../../context/layers/layer-data';
import { layerDataSelector } from '../../../context/mapStateSlice/selectors';
import { Extent } from '../Layers/raster-utils';
import { availableDatesSelector } from '../../../context/serverStateSlice';
import {
  AnalysisDispatchParams,
  analysisResultSelector,
  clearAnalysisResult,
  isAnalysisLayerActiveSelector,
  isAnalysisLoadingSelector,
  requestAndStoreAnalysis,
  setIsMapLayerActive,
} from '../../../context/analysisResultStateSlice';
import AnalysisTable from './AnalysisTable';
import {
  getAnalysisTableColumns,
  downloadCSVFromTableData,
} from '../../../utils/analysis-utils';
import {
  useAnalyserReducer,
  AnalyserForm,
  URLParamList,
} from './AnalyserReducer';
import LayerDropdown from '../Layers/LayerDropdown';

const boundaryLayer = getBoundaryLayerSingleton();

function Analyser({ classes }: AnalyserProps) {
  const dispatch = useDispatch();
  const history = useHistory();
  const boundaryLayerData = useSelector(layerDataSelector(boundaryLayer.id)) as
    | LayerData<BoundaryLayerProps>
    | undefined;

  const availableDates = useSelector(availableDatesSelector);
  const analysisResult = useSelector(analysisResultSelector);

  const isAnalysisLoading = useSelector(isAnalysisLoadingSelector);
  const isMapLayerActive = useSelector(isAnalysisLayerActiveSelector);

  const [isAnalyserFormOpen, setIsAnalyserFormOpen] = useState(false);
  const [isTableViewOpen, setIsTableViewOpen] = useState(true);
  const activeUrl: React.MutableRefObject<string> = useRef<string>('');

  // form elements
  const [form, setForm] = useAnalyserReducer();

  // set default date after dates finish loading and when hazard layer changes
  useEffect(() => {
    const dates = form.hazardLayerId
      ? availableDates[
          (LayerDefinitions[form.hazardLayerId] as WMSLayerProps)
            ?.serverLayerName
        ]
      : null;
    if (!dates || dates.length === 0) {
      setForm({
        type: 'SET_SELECTED_DATE',
        payload: {},
      });
    } else {
      setForm({
        type: 'SET_SELECTED_DATE',
        payload: { date: dates[dates.length - 1] },
      });
    }
  }, [availableDates, form.hazardLayerId, setForm]);

  const adminBoundariesExtent = useMemo(() => {
    if (!boundaryLayerData) {
      // not loaded yet. Should be loaded in MapView
      return null;
    }
    return bbox(boundaryLayerData.data) as Extent; // we get extents of admin boundaries to give to the api.
  }, [boundaryLayerData]);

  const statisticOptions = Object.entries(AggregationOperations).map(stat => (
    <FormControlLabel
      key={stat[0]}
      value={stat[1]}
      control={
        <Radio className={classes.radioOptions} color="default" size="small" />
      }
      label={stat[0]}
    />
  ));

  const clearAnalysis = () => {
    dispatch(clearAnalysisResult());

    // Empty previous set data.
    setForm({ type: 'CLEAR_FORM' });

    // Reset URL in browser.
    const URLStringToReplace: string = removePropsFromURL(
      history.location.search,
      URLParamList,
    );
    history.replace(`?${URLStringToReplace}`);
  };

  const runAnalyser = async () => {
    if (!adminBoundariesExtent) {
      return;
    } // hasn't been calculated yet

    // Detect from issues and throw errors.
    if (!form.selectedDate) {
      throw new Error('Date must be given to run analysis');
    }

    if (!form.hazardLayerId || !form.baselineLayerId) {
      throw new Error('Layer should be selected to run analysis');
    }

    // Build URL string.
    const analyserShareURL: string = Object.keys(form)
      .map((prop: string) => {
        return [prop, form[prop as keyof AnalyserForm]]
          .map(param =>
            param && param !== '' ? encodeURIComponent(param) : undefined,
          )
          .filter(Boolean)
          .join('=');
      })
      .join('&');

    // Make sure to run the nalyser only when needed.
    if (activeUrl.current !== analyserShareURL) {
      // Store the new URL.
      activeUrl.current = analyserShareURL;

      const selectedHazardLayer = LayerDefinitions[
        form.hazardLayerId
      ] as WMSLayerProps;
      const selectedBaselineLayer = LayerDefinitions[
        form.baselineLayerId
      ] as NSOLayerProps;

      const params: AnalysisDispatchParams = {
        hazardLayer: selectedHazardLayer,
        baselineLayer: selectedBaselineLayer,
        date: form.selectedDate,
        statistic: form.statistic as AggregationOperations,
        extent: adminBoundariesExtent,
        threshold: {
          above: form.aboveThreshold
            ? parseFloat(form.aboveThreshold as string)
            : undefined,
          below: form.belowThreshold
            ? parseFloat(form.belowThreshold as string)
            : undefined,
        },
      };

      await dispatch(requestAndStoreAnalysis(params));

      // Set share URL into the browser.
      history.replace(`?share=true&${analyserShareURL}`);
    }
  };

  const isFormInvalid = (): boolean => {
    return (
      !!form.thresholdError || // if there is a threshold error
      !form.selectedDate || // or date hasn't been selected
      !form.hazardLayerId || // or hazard layer hasn't been selected
      !form.baselineLayerId // or baseline layer hasn't been selected
    );
  };

  useEffect(() => {
    // Early return if data not loaded.
    if (!boundaryLayerData || !availableDates) {
      return;
    }

    const URLParams: propsFromURL = extractPropsFromURL(
      history.location.search,
      URLParamList,
    );

    // Set data from the URL.
    if (URLParams.fromURL) {
      // Set all from data from the URL.
      setForm({
        type: 'SET_FORM',
        payload: { params: URLParams.params },
      });

      // Avoid Running Analyser if required data are invalid or it's not a share link.
      if (!isFormInvalid() && URLParams.fromURL) {
        runAnalyser();
      }
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [
    boundaryLayerData,
    availableDates,
    history.location.search,
    form.selectedDate,
  ]);

  return (
    <div className={classes.analyser}>
      <Button
        variant="contained"
        color="primary"
        onClick={() => {
          setIsAnalyserFormOpen(!isAnalyserFormOpen);
        }}
      >
        <BarChart fontSize="small" />
        <Typography variant="body2" className={classes.analyserLabel}>
          Run Analysis
        </Typography>
        <ArrowDropDown fontSize="small" />
      </Button>

      <Box
        className={classes.analyserMenu}
        width={isAnalyserFormOpen ? 'min-content' : 0}
        padding={isAnalyserFormOpen ? '10px' : 0}
      >
        {isAnalyserFormOpen ? (
          <div>
            <div className={classes.newAnalyserContainer}>
              <div className={classes.analyserOptions}>
                <Typography variant="body2">Hazard Layer</Typography>
                <LayerDropdown
                  type="wms"
                  value={form.hazardLayerId}
                  setValue={(paramLayerKey: LayerKey) => {
                    setForm({
                      type: 'SET_LAYER_ID',
                      payload: { type: 'hazard', layerKey: paramLayerKey },
                    });
                  }}
                  title="Hazard Layer"
                  className={classes.selector}
                  placeholder="Choose hazard layer"
                />
              </div>
              <div className={classes.analyserOptions}>
                <Typography variant="body2">Statistic</Typography>
                <FormControl component="div">
                  <RadioGroup
                    name="statistics"
                    value={form.statistic}
                    onChange={(
                      paramEvent: React.ChangeEvent<HTMLInputElement>,
                    ) => {
                      setForm({
                        type: 'SET_STATISTIC',
                        payload: {
                          statistic: paramEvent.target
                            .value as AggregationOperations,
                        },
                      });
                    }}
                    row
                  >
                    {statisticOptions}
                  </RadioGroup>
                </FormControl>
              </div>
              <div className={classes.analyserOptions}>
                <Typography variant="body2">Baseline Layer</Typography>
                <LayerDropdown
                  type="nso"
                  value={form.baselineLayerId}
                  setValue={(paramLayerKey: LayerKey) => {
                    setForm({
                      type: 'SET_LAYER_ID',
                      payload: { type: 'baseline', layerKey: paramLayerKey },
                    });
                  }}
                  title="Baseline Layer"
                  className={classes.selector}
                  placeholder="Choose baseline layer"
                />
              </div>
              <div className={classes.analyserOptions}>
                <Typography variant="body2">Threshold</Typography>
                <TextField
                  id="filled-number"
                  error={!!form.thresholdError}
                  helperText={form.thresholdError}
                  className={classes.numberField}
                  label="Min"
                  type="number"
                  value={form.aboveThreshold}
                  onChange={(
                    paramEvent: React.ChangeEvent<HTMLInputElement>,
                  ) => {
                    setForm({
                      type: 'SET_THRESHOLD',
                      payload: {
                        type: 'above',
                        value: paramEvent.target.value,
                      },
                    });
                  }}
                  variant="filled"
                />
                <TextField
                  id="filled-number"
                  label="Max"
                  className={classes.numberField}
                  value={form.belowThreshold}
                  onChange={(
                    paramEvent: React.ChangeEvent<HTMLInputElement>,
                  ) => {
                    setForm({
                      type: 'SET_THRESHOLD',
                      payload: {
                        type: 'below',
                        value: paramEvent.target.value,
                      },
                    });
                  }}
                  type="number"
                  variant="filled"
                />
              </div>
              <div className={classes.analyserOptions}>
                <Typography variant="body2">Date</Typography>
                <DatePicker
                  selected={
                    form.selectedDate ? new Date(form.selectedDate) : null
                  }
                  onChange={date => {
                    setForm({
                      type: 'SET_SELECTED_DATE',
                      payload: { date: date?.getTime() || form.selectedDate },
                    });
                  }}
                  maxDate={new Date()}
                  todayButton="Today"
                  peekNextMonth
                  showMonthDropdown
                  showYearDropdown
                  dropdownMode="select"
                  customInput={<Input />}
                  popperClassName={classes.calendarPopper}
                  includeDates={
                    form.hazardLayerId
                      ? availableDates[
                          (LayerDefinitions[
                            form.hazardLayerId
                          ] as WMSLayerProps).serverLayerName
                        ]?.map(d => new Date(d)) || []
                      : []
                  }
                />
              </div>
            </div>

            {!isAnalysisLoading && analysisResult && (
              <>
                <FormGroup>
                  <FormControlLabel
                    control={
                      <Switch
                        color="default"
                        checked={isMapLayerActive}
                        onChange={e =>
                          dispatch(setIsMapLayerActive(e.target.checked))
                        }
                      />
                    }
                    label="Map View"
                  />
                  <FormControlLabel
                    control={
                      <Switch
                        color="default"
                        checked={isTableViewOpen}
                        onChange={e => setIsTableViewOpen(e.target.checked)}
                      />
                    }
                    label="Table View"
                  />
                </FormGroup>
                {isTableViewOpen && (
                  <AnalysisTable
                    tableData={analysisResult.tableData}
                    columns={getAnalysisTableColumns(analysisResult)}
                  />
                )}
                <Button
                  className={classes.innerAnalysisButton}
                  onClick={() => downloadCSVFromTableData(analysisResult)}
                >
                  <Typography variant="body2">Download</Typography>
                </Button>
                <Button
                  className={classes.innerAnalysisButton}
                  onClick={clearAnalysis}
                >
                  <Typography variant="body2">Clear Analysis</Typography>
                </Button>
              </>
            )}
            {!analysisResult && (
              <Button
                className={classes.innerAnalysisButton}
                onClick={runAnalyser}
                disabled={
                  isFormInvalid() || isAnalysisLoading // or analysis is currently loading
                }
              >
                <Typography variant="body2">Run Analysis</Typography>
              </Button>
            )}
            {isAnalysisLoading ? <LinearProgress /> : null}
          </div>
        ) : null}
      </Box>
    </div>
  );
}

const styles = (theme: Theme) =>
  createStyles({
    analyser: {
      zIndex: theme.zIndex.drawer,
      textAlign: 'left',
    },
    analyserLabel: {
      marginLeft: '10px',
    },
    analyserMenu: {
      backgroundColor: '#5A686C',
      maxWidth: '100vw',
      color: 'white',
      overflowX: 'hidden',
      whiteSpace: 'nowrap',
      borderTopRightRadius: '10px',
      borderBottomRightRadius: '10px',
      height: 'auto',
      maxHeight: '60vh',
    },
    analyserButton: {
      height: '36px',
      'margin-left': '3px',
    },
    analyserOptions: {
      padding: '5px 0px',
    },
    newAnalyserContainer: {
      padding: '5px',
      marginTop: '10px',
    },
    radioOptions: {
      '&.Mui-checked': { color: grey[50] },
      padding: '2px 10px 2px 20px',
    },
    innerAnalysisButton: {
      backgroundColor: '#3d474a',
      margin: '10px',
      '&.Mui-disabled': { opacity: 0.5 },
    },
    selectorLabel: {
      '&.Mui-focused': { color: 'white' },
    },
    selector: {
      margin: '5px',
    },
    numberField: {
      paddingLeft: '10px',
      marginTop: '10px',
      width: '85.5px',
      '& .Mui-focused': { color: 'white' },
    },
    calendarPopper: {
      zIndex: 3,
    },
  });

interface AnalyserProps extends WithStyles<typeof styles> {}

export default withStyles(styles)(Analyser);
