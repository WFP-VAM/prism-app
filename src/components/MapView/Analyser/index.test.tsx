import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';

import Analyser from '.';
import { store } from '../../../context/store';

jest.mock('../Layers/LayerDropdown', () => 'mock-Layer-Dropdown');

test('renders as expected', () => {
  const rendered = render(
    <Provider store={store}>
      <Router>
        <Analyser />
      </Router>
    </Provider>,
  );
  return rendered
    .findByText('Run Analysis')
    .then(btn => btn.click()) // open analyser menu (default closed)
    .then(() => expect(rendered.container).toMatchSnapshot());
});
