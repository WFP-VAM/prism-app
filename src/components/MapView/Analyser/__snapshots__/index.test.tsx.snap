// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="Analyser-analyser-1"
  >
    <mock-button
      color="primary"
      variant="contained"
    >
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-chart-bar fa-w-16 "
        data-icon="chart-bar"
        data-prefix="fas"
        focusable="false"
        role="img"
        style="margin-right: 10px; font-size: 1.6em;"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M332.8 320h38.4c6.4 0 12.8-6.4 12.8-12.8V172.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v134.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h38.4c6.4 0 12.8-6.4 12.8-12.8V76.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v230.4c0 6.4 6.4 12.8 12.8 12.8zm-288 0h38.4c6.4 0 12.8-6.4 12.8-12.8v-70.4c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v70.4c0 6.4 6.4 12.8 12.8 12.8zm96 0h38.4c6.4 0 12.8-6.4 12.8-12.8V108.8c0-6.4-6.4-12.8-12.8-12.8h-38.4c-6.4 0-12.8 6.4-12.8 12.8v198.4c0 6.4 6.4 12.8 12.8 12.8zM496 384H64V80c0-8.84-7.16-16-16-16H16C7.16 64 0 71.16 0 80v336c0 17.67 14.33 32 32 32h464c8.84 0 16-7.16 16-16v-32c0-8.84-7.16-16-16-16z"
          fill="currentColor"
        />
      </svg>
      <mock-typography
        variant="body2"
      >
        Run Analysis
      </mock-typography>
      <svg
        aria-hidden="true"
        class="svg-inline--fa fa-caret-down fa-w-10 "
        data-icon="caret-down"
        data-prefix="fas"
        focusable="false"
        role="img"
        style="margin-left: 10px;"
        viewBox="0 0 320 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          fill="currentColor"
        />
      </svg>
    </mock-button>
    <div
      class="MuiBox-root MuiBox-root-12 Analyser-analyserMenu-2"
    >
      <div>
        <div
          class="Analyser-newAnalyserContainer-5"
        >
          <div
            class="Analyser-analyserOptions-4"
          >
            <mock-typography
              variant="body2"
            >
              Hazard Layer
            </mock-typography>
            <mock-layer-dropdown
              classname="Analyser-selector-9"
              placeholder="Choose hazard layer"
              title="Hazard Layer"
              type="wms"
            />
          </div>
          <div
            class="Analyser-analyserOptions-4"
          >
            <mock-typography
              variant="body2"
            >
              Statistic
            </mock-typography>
            <div
              class="MuiFormControl-root"
            >
              <div
                class="MuiFormGroup-root MuiFormGroup-row"
                role="radiogroup"
              >
                <label
                  class="MuiFormControlLabel-root"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-13 MuiRadio-root Analyser-radioOptions-6 PrivateSwitchBase-checked-14 Mui-checked"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        checked=""
                        class="PrivateSwitchBase-input-16"
                        name="statistics"
                        type="radio"
                        value="mean"
                      />
                      <div
                        class="PrivateRadioButtonIcon-root-17 PrivateRadioButtonIcon-checked-19"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                          />
                        </svg>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-18 MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                          />
                        </svg>
                      </div>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </span>
                  <mock-typography
                    classname="MuiFormControlLabel-label"
                    component="span"
                  >
                    Mean
                  </mock-typography>
                </label>
                <label
                  class="MuiFormControlLabel-root"
                >
                  <span
                    aria-disabled="false"
                    class="MuiButtonBase-root MuiIconButton-root PrivateSwitchBase-root-13 MuiRadio-root Analyser-radioOptions-6"
                  >
                    <span
                      class="MuiIconButton-label"
                    >
                      <input
                        class="PrivateSwitchBase-input-16"
                        name="statistics"
                        type="radio"
                        value="median"
                      />
                      <div
                        class="PrivateRadioButtonIcon-root-17"
                      >
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
                          />
                        </svg>
                        <svg
                          aria-hidden="true"
                          class="MuiSvgIcon-root PrivateRadioButtonIcon-layer-18 MuiSvgIcon-fontSizeSmall"
                          focusable="false"
                          viewBox="0 0 24 24"
                        >
                          <path
                            d="M8.465 8.465C9.37 7.56 10.62 7 12 7C14.76 7 17 9.24 17 12C17 13.38 16.44 14.63 15.535 15.535C14.63 16.44 13.38 17 12 17C9.24 17 7 14.76 7 12C7 10.62 7.56 9.37 8.465 8.465Z"
                          />
                        </svg>
                      </div>
                    </span>
                    <span
                      class="MuiTouchRipple-root"
                    />
                  </span>
                  <mock-typography
                    classname="MuiFormControlLabel-label"
                    component="span"
                  >
                    Median
                  </mock-typography>
                </label>
              </div>
            </div>
          </div>
          <div
            class="Analyser-analyserOptions-4"
          >
            <mock-typography
              variant="body2"
            >
              Baseline Layer
            </mock-typography>
            <mock-layer-dropdown
              classname="Analyser-selector-9"
              placeholder="Choose baseline layer"
              title="Baseline Layer"
              type="nso"
            />
          </div>
          <div
            class="Analyser-analyserOptions-4"
          >
            <mock-typography
              variant="body2"
            >
              Threshold
            </mock-typography>
            <mock-textfield
              classname="Analyser-numberField-10"
              error="false"
              id="filled-number"
              label="Min"
              type="number"
              value=""
              variant="filled"
            />
            <mock-textfield
              classname="Analyser-numberField-10"
              id="filled-number"
              label="Max"
              type="number"
              value=""
              variant="filled"
            />
          </div>
          <div
            class="Analyser-analyserOptions-4"
          >
            <mock-typography
              variant="body2"
            >
              Date
            </mock-typography>
            <div
              class="react-datepicker-wrapper"
            >
              <div
                class="react-datepicker__input-container"
              >
                <div
                  class="MuiInputBase-root MuiInput-root MuiInput-underline"
                >
                  <input
                    class="MuiInputBase-input MuiInput-input"
                    type="text"
                    value=""
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <mock-button
          classname="Analyser-innerAnalysisButton-7"
          disabled="true"
        >
          <mock-typography
            variant="body2"
          >
            Run Analysis
          </mock-typography>
        </mock-button>
      </div>
    </div>
  </div>
</div>
`;
